### PR TITLE
Cows! Cows everywhere!

### LINKED ISSUE #114: Store borrowed fields using `Cow`
The current implementation stores `&'a T` for many fields (I'm particularly thinking of [`ResponseCode`](https://docs.rs/imap-proto/0.12/imap_proto/types/enum.ResponseCode.html), but it applies to others). That makes a lot of sense for parsing, but makes it hard to clone the type to keep it beyond the parsing step.

This could be mitigated by replacing every `&'a T` with `Cow<'a, T>` and then adding
```rust
impl<'a> ResponseCode<'a> {
   pub fn into_owned(self) -> ResponseCode<'static> { }
}
```
to each such type, which just maps all the internal `Cow`'s into their owned variants.

That would make these types a lot more ergonomic to use in more contexts, and would avoid forcing downstream libraries to keep their own owned versions of these types!

### PR DESCRIPTION
**Pull request by jonhoo**
Fixes #114.

### CODE REVIEW

File: imap-proto/src/types.rs [hunk 1]
```diff
@@ -88,6 +89,37 @@ impl From<u32> for UidSetMember {
...
 }
 
+impl<'a> ResponseCode<'a> {
+    pub fn into_owned(self) -> ResponseCode<'static> {
+        match self {
+            ResponseCode::Alert => ResponseCode::Alert,
+            ResponseCode::BadCharset(v) => ResponseCode::BadCharset(v.map(|vs| {
+                vs.into_iter()
+                    .map(Cow::into_owned)
+                    .map(Cow::Owned)
```
Comment (djc) [inline imap-proto/src/types.rs]: Surely we we can do better than `map().map()`? Maybe `map(|v| Cow::into_owned(v).into())`?

Comment (jonhoo) [reply imap-proto/src/types.rs]: I actually found the former a little nicer to read since it doesn't require the closure, but will change :+1:

Comment (jonhoo) [reply imap-proto/src/types.rs]: Added a helper function — much nicer now!

**Comment by jonhoo** [comment]
:tada: Do you have an estimate for when the next (breaking) release might happen? I'd like to include this change in the upcoming `imap 3.0.0` release if possible.

**Comment by djc** [comment]
I expect that to happen about 5 seconds ago. 👍

**Pull request merged**